Draft: Refactor Edition Workspace

olivierSaintCyr opened the pull request:
**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines:https://github.com/infra-geo-ouverte/igo2/blob/master/.github/CONTRIBUTING.md#git-commit-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What is the current behavior?** (You can also link to an open issue here)



**What is the new behavior?**
A refactor of the edition workspace logic #1667


**Does this PR introduce a breaking change?** (check one with "x")
- [x] Yes
- [ ] No

**If this PR contains a breaking change, please describe the impact and migration path for existing applications:**
Remove of the send attribute in EditionOptions due to redundancy with readonly attribute

**Other information**:

olivierSaintCyr committed:
init new edition workspace.

olivierSaintCyr committed:
feat(edition): create table template factory

olivierSaintCyr committed:
feat(edition); add workspace to layer

olivierSaintCyr committed:
add edition-workspace to edition actions

olivierSaintCyr committed:
fix(edition-workspace): not in resolution

olivierSaintCyr committed:
refactor(edition-workspace): inResolutionRange

olivierSaintCyr committed:
feat(edition-workspace): impl table edition

olivierSaintCyr committed:
feat(edition workspace): cancel edition

olivierSaintCyr committed:
refactor(edition-workspace): table template factory

olivierSaintCyr committed:
feat(edition-workspace): edit feature attributes

olivierSaintCyr committed:
fix(edition): loading not updating

olivierSaintCyr committed:
feat(rest-api-edition): impl update body

olivierSaintCyr committed:
feat(edition-workspace): delete

olivierSaintCyr committed:
feat(edition-workspace): create feature edition & cancel

olivierSaintCyr committed:
feat(edition-workspace): create feature without geometry

olivierSaintCyr committed:
adapt edtion

olivierSaintCyr committed:
fix build

olivierSaintCyr committed:
fix build

olivierSaintCyr committed:
fix(edition-workspace): add when edit not canceling edit

olivierSaintCyr committed:
fix(edition-workspace): update not sending data

olivierSaintCyr committed:
feat(edition-workspace): message when success and error

olivierSaintCyr committed:
Merge branch 'next' into f/edition-wfs-ogcapi-datasource

olivierSaintCyr committed:
refactor(edition-workspace): feature creation

olivierSaintCyr committed:
feat(edition-workspace): enable disable geo edition

olivierSaintCyr committed:
feat(workspace-editor): add point on map

alecarn reviewed (changes requested):
Un premier tour de revue pour me familiariser avec le code, c'est bien partie👍 
Commentaire générale, améliorer le typage pour solidifier nos bases

olivierSaintCyr committed:
refactor(edition-workspace): add Feature type to Workspace

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/new-edition-workspace/edition-table-template-factory.ts:
```diff
@@ -0,0 +1,194 @@
+        first()
+      )
+      .subscribe((entities) => {
+        const ol = (entities[0] as Feature).ol as olFeature<OlGeometry>;
```
Bonne idée! Je le change tout de suite!

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/new-edition-workspace/edition-table-template-factory.ts:
```diff
@@ -0,0 +1,194 @@
+      }
+    };
+
+    workspace.isLoading$.subscribe((isLoading) => {
+      confirmButton.disabled = isLoading;
+      cancelButton.disabled = isLoading;
+    });
```
Oui j'aurais aimer ça que nous aurons pus projeter les changement plus profond. Avant on utilisais un object boolean loading qui était insérer dans le template pour activé et desactiver les boutons. Par contre, ceci entrainais un bug. Le bug était que les boutons ne se disable pas lorsque que loading est set a true vu que l'interpreteur js avait copier la valeur de l'attribut loading et non la référence vu qu'un boolean est un object primitif. Alors il fallait trouver un moyen d'aller update le template directement afin que les changement se propage dans le UI. C'est pourquoi j'ai décider de subscribe sur l'observable isLoading$ afin de découpler le workspace du table template et laisser la responsabilité au template factory de setup le template!

olivierSaintCyr committed:
fix(layer-workspace): add Feature type to all layer-workspace

olivierSaintCyr committed:
refactor(new-edition-workspace): limit export scope to only new-edition-workspace

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/new-edition-workspace/edition-workspace-factory.service.ts:
```diff
@@ -0,0 +1,116 @@
+import { RestAPIEdition } from './rest-api-edition';
+
+@Injectable({
+  providedIn: 'root'
```
Nous avons pas besoin de l'exposer globallement nice catch j'ai limiter le export uniquement au new-edition-workspace

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/new-edition-workspace/edition-table-template-factory.ts:
```diff
@@ -0,0 +1,194 @@
+  private getButtonTemplate(
+    workspace: NewEditionWorkspace,
+    layer: VectorLayer
+  ) {
```
Jai typer les columns par contre certains attributs ne sont pas dans l'interface EntityTableColumn et en les retirant le UI silent fail. Malheuresement je pense pas avoir le temps dans mon mandat pour le fix. Je vais rajouter un TODO pour pas qu'on oublie de le faire!

olivierSaintCyr replied:
voir commentaire plus haut!

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/new-edition-workspace/edition-table-template-factory.ts:
```diff
@@ -0,0 +1,194 @@
+    const relations = layer.dataSource.options.relations ?? [];
+    const relationsColumns = this.getRelationsTemplate(relations);
+
+    const columns: unknown[] = sourceFieldsColumns;
```
Voir le commentaire plus haut!

olivierSaintCyr committed:
refactor(edition-ws-table-template-factory): add typing todo

olivierSaintCyr committed:
refactor(EditionOptions)!: add values for modifyMethod

BREAKING CHANGE: rename modifyProtocol to modifyMethod
BREAKING CHANGE: modifyMethod can only be post and patch
BREAKING CHANGE: remove send attribute of EditionOptions

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/new-edition-workspace/new-edition-workspace.ts:
```diff
@@ -0,0 +1,348 @@
+    const headers = new HttpHeaders(modifyHeaders);
+
+    this.isLoadingSubject.next(true);
+    this.http[modifyMethod ?? 'patch'](url, this.getUpdateBody(feature), {
```
Done! Jai modifier les valeur de modifyMethod pour 'patch' | 'post'

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/new-edition-workspace/new-edition-workspace.ts:
```diff
@@ -0,0 +1,348 @@
+        this.isLoadingSubject.next(false);
+        this.handleEditionError(error);
+      }
+    });
```
Personnellement je pense que oui car la classe qui utilise ces methods est le EditionWorkspaceFactory et lui a uniquement une tache créer les edition-workspace . Nous pourrions créé un service qui wrap les workspaces et call les methods add modify etc. ce qui va potentiellement augmenter la complexité. Bref, TBD

olivierSaintCyr committed:
refactor(layer-workspace): rename layer-workspace to any-workspace

olivierSaintCyr commented on packages/geo/src/lib/workspace/shared/layer-workspace.ts:
```diff
@@ -0,0 +1,12 @@
+  WfsWorkspace
+} from './';
+
+export type LayerWorkspace =
```
Done!
